Name and unpair individual clients

xanderfrangos opened the pull request:
## Description
Introduces the option to name clients during PIN pairing, along with a UI to remove individual client certificates. The format of `sunshine_state.json` was updated to include a `named_certs` array instead of the previous `certs` array. The old array is used for migration to the new format on startup.

Here's an example of the new format:

```json
{
    ...
    "root": {
        "uniqueid": "2258BFE6-698A-422C-5F0F-55B49A69C69E",
        "devices": [
            {
                "uniqueid": "0123456789ABCDEF",
                "named_certs": [
                    {
                        "name": "NVIDIA Shield",
                        "cert": "-----BEGIN CERTIFICATE-----\nCERT\n-----END CERTIFICATE-----\n",
                        "uniqueid": "41310952-60C8-1B01-A615-033CC8275E5F"
                    },
                    {
                        "name": "Steam Deck",
                        "cert": "-----BEGIN CERTIFICATE-----\nCERT\n-----END CERTIFICATE-----\n",
                        "uniqueid": "AE113204-32D9-6607-6285-7F3C200440EB"
                    }
                ]
            }
        ]
    }
}

```

The `named_certs` array includes a user-defined name and a generated UUID. The UUID allows for the same name to be used multiple times. It also helps target the correct client over the API without exposing the certificate.

The UI for managing clients has been added to the PIN page, under the PIN input. If a client wasn't named during the pairing process, or the certificate was imported from the old format, the client will be listed as "Unknown". After removing a paired client from the new UI, the user will be prompted to restart Sunshine in order to finish removing the certificate. 

While working on this, I was made aware of the [TheElixZammuto/Sunshine-1/state-revamp](https://github.com/TheElixZammuto/Sunshine-1/tree/state-revamp) fork, which has a similar goal. However, it looks like it's still missing the UI, API, and server code to deal with "unpairing". It also hasn't been updated since September, so it can't merge cleanly as-is. I understand if you prefer to use what that fork over mine, but I figured I would clean up my code and submit it, in case it's useful 

### Screenshot
![Screenshot 2024-01-20 032346](https://github.com/LizardByte/Sunshine/assets/33106561/12364477-c0e2-481e-9a83-3d52760ebf43)



### Issues Fixed or Closed
https://discord.com/channels/804382334370578482/1197647357030965328


## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency update (updates to dependencies)
- [ ] Documentation update (changes to documentation)
- [ ] Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

## Branch Updates
LizardByte requires that branches be up-to-date before merging. This means that after any PR is merged, this branch
must be updated before it can be merged. You must also
[Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I want maintainers to keep my branch updated

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/LizardByte/Sunshine?pullRequest=2042) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/LizardByte/Sunshine?pullRequest=2042) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/LizardByte/Sunshine?pullRequest=2042) it.</sub>

ReenigneArcher commented:
@TheElixZammuto any thoughts on using this over what were working on?

TheElixZammuto commented:
> @TheElixZammuto any thoughts on using this over what were working on?

Just two things:
1. I would rename uniqueid to uuid or a similar name, since uniqueid is a protocol-specific name which can lead to confusion (it's not the same uniqueid as the one the moonlight client sends)
2. I would prefer to have named_certs as a root element instead of inside "devices", which is useless right now since Moonlight always send the same uniqueid

xanderfrangos commented:
@TheElixZammuto Easy enough. 👍 Regarding setting it as a "root element", do you mean the root of the JSON object or the `root` property (that currently contains `uniqueid` and `devices`)?

TheElixZammuto commented:
My example idea
```
{
    "root": {
        "uniqueid": "2258BFE6-698A-422C-5F0F-55B49A69C69E",
        "named_devices": [
            {
                "name": "NVIDIA Shield",
                "cert": "-----BEGIN CERTIFICATE-----\nCERT\n-----END CERTIFICATE-----\n",
                "uuid": "41310952-60C8-1B01-A615-033CC8275E5F"
            },
            {
                "name": "Steam Deck",
                "cert": "-----BEGIN CERTIFICATE-----\nCERT\n-----END CERTIFICATE-----\n",
                "uuid": "AE113204-32D9-6607-6285-7F3C200440EB"
            }
        ]
    }
}

```

xanderfrangos commented:
@TheElixZammuto I've updated `sunshine_state.json` to use the format you've outlined. I've also refactored `nvhttp.cpp` to no longer map `client_t`s to `uniqueid`s, since `root.devices` is gone (and it's always Moonlight's "0123456789ABCDEF" anyway). Only a single `client_t` is used now.

ReenigneArcher reviewed (changes requested):
In my opinion the client list and the ability to remove them should be in the troubleshooting section, near the unpair all clients option.

The pin/pair page should be lightweight.

xanderfrangos commented:
@ReenigneArcher Moved to Troubleshooting 👍

![localhost_37990_troubleshooting](https://github.com/LizardByte/Sunshine/assets/33106561/0392239e-829c-4b28-8307-751ba9ce85e0)

ReenigneArcher commented on src_assets/common/assets/web/troubleshooting.html:
```diff
@@ -82,23 +82,38 @@ <h2 id="restart">Restart Sunshine</h2>
+            </button>
+          </div>
+          <div id="apply-alert" class="alert alert-success d-flex align-items-center mt-3" :style="{ 'display': (showApplyMessage ? 'flex !important': 'none !important') }">
+            <div class="me-2"><b>Success!</b> Click 'Apply' to restart Sunshine and apply changes. This will terminate any running sessions.</div>
```
Is it necessary to restart Sunshine? I'm guessing you're doing this to force disconnecting a client if it was one of those that were removed?

There's probably a way we can disconnect that single client instead of restarting Sunshine altogether.

xanderfrangos replied:
Two reasons:
1. Yes, to disconnect the client.
2. The certificate is unfortunately is still valid after unpairing a single client, so it can simply reconnect without issue. I'm not sure why. If someone can fix this, then only the session would need to end.

If access can be fully revoked and the client is disconnected, I think that would be a lot better than requesting a restart.

ReenigneArcher replied:
@cgutman do you what we can do to remove a single client without restarting sunshine?

xanderfrangos replied:
I think this is the only remaining concern.

ReenigneArcher replied:
I think we would have to remove it from `cert_chain`. In https://github.com/LizardByte/Sunshine/pull/2365/files I am removing all of them for the unpair all.

ReenigneArcher commented on src/confighttp.cpp:
```diff
@@ -745,6 +799,8 @@ namespace confighttp {
     server.resource["^/api/apps/([0-9]+)$"]["DELETE"] = deleteApp;
     server.resource["^/api/clients/unpair$"]["POST"] = unpairAll;
+    server.resource["^/api/clients/list$"]["GET"] = listClients;
+    server.resource["^/api/clients/unpair-single$"]["POST"] = unpairClient;
```
```suggestion
    server.resource["^/api/clients/unpair$"]["POST"] = unpair;
```

ReenigneArcher commented on src/confighttp.cpp:
```diff
@@ -700,6 +701,59 @@ namespace confighttp {
   }
 
+  void
+  unpairClient(resp_https_t response, req_https_t request) {
```
```suggestion
  unpair(resp_https_t response, req_https_t request) {
```

ReenigneArcher commented on src/confighttp.cpp:
```diff
@@ -700,6 +701,59 @@ namespace confighttp {
+      outputTree.put("status", nvhttp::unpair_client(uuid));
+    }
+    catch (std::exception &e) {
+      BOOST_LOG(warning) << "UnpairClient: "sv << e.what();
```
```suggestion
      BOOST_LOG(warning) << "Unpair: "sv << e.what();
```

ReenigneArcher commented on src_assets/common/assets/web/troubleshooting.html:
```diff
@@ -180,13 +198,38 @@ <h2 id="logs">Logs</h2>
           fetch("/api/clients/unpair", { method: "POST" })
```
```suggestion
          fetch("/api/clients/unpair-all", { method: "POST" })
```

ReenigneArcher commented on src/nvhttp.cpp:
```diff
@@ -217,48 +222,72 @@ namespace nvhttp {
-      client.uniqueID = uniqID;
+    auto root = tree.get_child("root");
+    client_t client;
+    client.uniqueID = "0123456789ABCDEF"s;
```
Why is this hard coded?

xanderfrangos replied:
I wasn't sure if the `uniqueID` is being used elsewhere in Sunshine. I had refactored out most of the code related to mapping certificates to `uniqueID`s, but left this (hard-coded to Moonlight's) just in case. If it's not needed, then I believe `uniqueID` can be removed from `client_t` entirely.

ReenigneArcher replied:
@cgutman pinging you since you're the most familiar with this.

cgutman replied:
You can remove it. Nvidia used it to identify separate clients, but it caused more issues than it solved (preventing a game started on one device from being resumed on another). That's why Moonlight just hardcodes a single value everywhere.

xanderfrangos replied:
Removed 👍 (https://github.com/LizardByte/Sunshine/pull/2042/commits/5ff80ada0b93dd8527c0c7e20e42e9f75065b0f3)

ReenigneArcher commented on src/nvhttp.cpp:
```diff
@@ -1096,7 +1150,47 @@ namespace nvhttp {
   erase_all_clients() {
-    map_id_client.clear();
+    client_t client;
+    client.uniqueID = "0123456789ABCDEF"s;
```
Same question... why hard coded?

xanderfrangos replied:
See previous response. It might be unnecessary.

xanderfrangos replied:
Removed 👍 (https://github.com/LizardByte/Sunshine/pull/2042/commits/5ff80ada0b93dd8527c0c7e20e42e9f75065b0f3)

xanderfrangos commented on src_assets/common/assets/web/pin.html:
```diff
@@ -8,10 +8,11 @@
       <div class="card flex-column d-flex p-4 mb-4">
-        <input type="text" pattern="\d*"  placeholder="PIN" id="pin-input" class="form-control my-4" />
+        <input type="number" pattern="\d*"  placeholder="PIN" id="pin-input" class="form-control mt-2" />
+        <input type="text" placeholder="Name (optional)" id="name-input" class="form-control my-4" />
```
Your call. It wasn't required before, so I didn't want to enforce it. If it's going to be required, I think the field name should better indicate that they're coming up with a name for the device. Maybe "Name this device" or "Set device name"?

ReenigneArcher replied:
I think it's important to have an identifiable name as user's aren't going to be able to differentiate based on a uuid.

Ideally we could also get the client type during the pair process, but I'm not sure that's possible at the moment.

I think "Device Name" is okay to use as the name of the field.

xanderfrangos replied:
It might also be nice to track/display when the client last connected and if there's an active session. But that seems like another PR for another day.

For now I'll make the field required and labeled "Device Name".

ReenigneArcher replied:
Agreed!

I'd like to see these items tracked in the future.

- Last known IP address
- Last known network type (LAN or WAN)
- Moonlight Client (e.g. Android, QT, iOS, etc.)
- Last known moonlight version
- Device Name (provided by Moonlight, not sure if feasible... but e.g. Nvidia Shield)
- Last session start and end times
- Last session application launched

cgutman replied:
I don't love the idea of making device name required. I'd much rather feed that info into the pairing process from the Moonlight side and allow the user to optionally adjust the name via this UI.

On the Moonlight side, it should be pretty easy to send the client's hostname across during the pairing process. We will run into some issues related to the privacy restrictions [on iOS](https://developer.apple.com/documentation/uikit/uidevice/1620015-name), but that's why we keep the optional field around as an override.

ReenigneArcher replied:
If Moonlight can send over enough details, then the name is probably not even necessary.

This is what Plex does. None of these fields can be modified, they are all collected from the client, and everything is pretty easy to identify.

![image](https://github.com/LizardByte/Sunshine/assets/42013603/cac0acf2-0ff9-4015-a4a2-2520990b361d)

The only issue I see is some of these can change after the pairing process. i.e. the client version (and IP address if we're going to track that somehow).

xanderfrangos replied:
It may still be helpful to optionally set a client name during pairing. 

I suspect most users don't name their devices at the OS level, which can make the Moonlight-provided name somewhat meaningless without more info. Tracking clients this way could be especially difficult if you're using Sunshine to share your PC with someone else (remote "local" co-op, for example). Personally, I would rather set the client name myself than explain to a friend that they need to set a name in Moonlight (or worse, their device's hostname).

Also, I'd expect (well, "hope" lol) that this PR will get approved before the required additions to Moonlight are implemented. So a name field is still needed (optional or not) for now.

ReenigneArcher commented:
@xanderfrangos sorry for the delays. Do you have time to resume this? I think it will fix #2305 which is probably a side effect of using a static uniqueID.

xanderfrangos commented:
@ReenigneArcher I unfortunately won't be able to work on this much until later this month. But to my understanding, the only outstanding concern (https://github.com/LizardByte/Sunshine/pull/2042#discussion_r1467967479) was something I wasn't sure how to help with anyway.

xanderfrangos commented:
@ReenigneArcher I've updated this branch to use the new localization library and (more importantly) utilize https://github.com/LizardByte/Sunshine/pull/2365. Individually unpaired devices can no longer start or resume a session. This should resolve https://github.com/LizardByte/Sunshine/pull/2042#discussion_r1467967479.

Since active sessions aren't automatically disconnected, I've added language to make that clearer for users. Here's an updated screenshot from the Troubleshooting page (taken after unpairing a device):

![Screenshot 2024-04-17 124543](https://github.com/LizardByte/Sunshine/assets/33106561/9341c251-cfec-458d-aac2-4ed33ec81c5d)

Also, the unpairing UI now consistently uses the word "device". I think that's a little more user-friendly than "client", but either is fine as long as it's consistent.

ReenigneArcher commented on src_assets/common/assets/web/public/assets/locale/en.json:
```diff
@@ -365,9 +367,12 @@
+    "unpair_all_desc": "Remove your paired devices. Unpaired devices with an active session will remain connected, but cannot start or resume a session.",
     "unpair_all_error": "Error while unpairing",
-    "unpair_all_success": "Unpair Successful!"
+    "unpair_all_success": "However, the device(s) may still be in an active session. Use the 'Force Close' button above to end any open sessions.",
```
```suggestion
    "unpair_all_success": "Unpair Successful! However, the device(s) may still be in an active session. Use the 'Force Close' button above to end any open sessions.",
```

ReenigneArcher commented on src_assets/common/assets/web/troubleshooting.html:
```diff
@@ -128,6 +140,8 @@ <h2 id="logs">{{ $t('troubleshooting.logs') }}</h2>
           unpairAllStatus: null,
           restartPressed: false,
+          showApplyMessage: false,
+          clients: [],
           logs: 'Loading...',
           logFilter: null,
           logInterval: null,
```
Can you alphabetize these? clients through logInterval

xanderfrangos replied:
Done 👍

I misunderstood what you meant initially, so the device list is now alphabetical too... which is better than being unsorted. 😅

ReenigneArcher commented on src_assets/common/assets/web/pin.html:
```diff
@@ -8,10 +8,11 @@
     <form action="" class="form d-flex flex-column align-items-center" id="form">
       <div class="card flex-column d-flex p-4 mb-4">
-        <input type="text" pattern="\d*" placeholder="PIN" id="pin-input" class="form-control my-4" />
+        <input type="text" pattern="\d*" placeholder="PIN" id="pin-input" class="form-control mt-2" required />
```
Can you update this placeholder to use the localized version? `navbar.pin` ... or move that one to `common`.

xanderfrangos replied:
Done 👍

ReenigneArcher commented on src/nvhttp.cpp:
```diff
@@ -1148,8 +1196,48 @@ namespace nvhttp {
    */
   void
   erase_all_clients() {
```
Maybe we should do a force close as part of the unpair all?

xanderfrangos replied:
Makes sense. "Unpair All" now force closes the open app/session. I've also updated the language to specify that the user only might need to do it themselves if unpairing individual devices.

ReenigneArcher replied:
Sorry, I am not seeing the change that would do the force close. Can you direct me to that part of the code?

xanderfrangos replied:
Sure: https://github.com/LizardByte/Sunshine/pull/2042/commits/1b6a828b1cdcda2854097b2d5788ee1d48f9dff9

ReenigneArcher commented on src_assets/common/assets/web/public/assets/locale/en.json:
```diff
@@ -324,7 +325,7 @@
     "home": "Home",
     "password": "Change Password",
-    "pin": "Pin",
+    "pin": "PIN",
```
```suggestion
    "pin": "Pin",
```

ReenigneArcher commented on src_assets/common/assets/web/public/assets/locale/en.json:
```diff
@@ -365,9 +367,13 @@
     "troubleshooting": "Troubleshooting",
     "unpair_all": "Unpair All",
-    "unpair_all_desc": "Remove all your paired devices",
+    "unpair_all_desc": "Remove your paired devices. Individually unpaired devices with an active session will remain connected, but cannot start or resume a session.",
```
```suggestion
    "unpair_desc": "Remove your paired devices. Individually unpaired devices with an active session will remain connected, but cannot start or resume a session.",
```

The key was throwing me off. If this is the generic description now for all of the unpairing options we should rename the key.

xanderfrangos replied:
Updated 👍 https://github.com/LizardByte/Sunshine/pull/2042/commits/52cb99611012dabdebb8caae51fb211bd8fdc58b

ReenigneArcher commented on src_assets/common/assets/web/troubleshooting.html:
```diff
@@ -75,24 +75,39 @@ <h2 id="restart">{{ $t('troubleshooting.restart_sunshine') }}</h2>
+            <div class="me-2"><b>{{ $t('_common.success') }}</b> {{ $t('troubleshooting.unpair_single_success') }}</div>
+            <button class="btn btn-success ms-auto apply" @click="clickedApplyBanner">{{ $t('_common.dismiss') }}</button>
+          </div>
+          <div class="alert alert-danger mt-3" v-if="unpairAllStatus === true">
```
```suggestion
          <div class="alert alert-success mt-3" v-if="unpairAllStatus === true">
```

ReenigneArcher commented:
@xanderfrangos sorry for the delays. This is the next one on our radar. Could you rebase it?

xanderfrangos commented:
@ReenigneArcher I've rebased locally, but need to test/check a few things first. Should have it up over the weekend.

ReenigneArcher commented:
@xanderfrangos thank you! Also, FYI, in the last 24 hours we changed our default branch back to `master`. Basically `nightly` was renamed to `master`. If you have any problems rebasing, that could be the reason.

xanderfrangos commented:
@ReenigneArcher It got a little messy, so I squashed my changes and rebased onto `master`. The only new change is fixing localization strings on the PIN page. Everything is now up-to-date and working on my end.

xanderfrangos commented:
On a side note: I know you're intentionally using "Pin", but "PIN" is what it should be. They have two different meanings. A "pin" is a small, narrow, sharp object (or the act of "pinning" something in place). A "PIN" is a "personal identification number". 

If you're trying to avoid "PIN" in all caps for some reason, you could use "pairing code"/"pairing" instead. But Moonlight already refers to it as a "PIN", so using any other language could confuse users.

ReenigneArcher commented:
@xanderfrangos Thanks! This should get merged soon unless we find any last minute things.

I find it easier to try to just have 1 commit and amend it when there are changes. It makes it a lot easier to rebase.

Regarding the PIN, that sounds right. We should probably address in a separate PR.

ReenigneArcher commented:
I don't know if there's a way to handle this without changes to Moonlight, but I did face this small issue.

When starting my testing initially, there was one unnamed device paired. I removed the Sunshine host from the client, and then repaired it. In Sunshine this showed 2 devices, one named and one unnamed.

When I removed the named one, it still had access since it was also the unnamed one.

Once, I removed the unnamed device on Sunshine's side, it was not able to connect.

Otherwise this looks good. I guess we can just recommend that people unpair all if they have any unnamed devices.

ReenigneArcher commented on src_assets/common/assets/web/pin.html:
```diff
@@ -24,7 +25,7 @@ <h1 class="my-4">{{ $t('pin.pin_pairing') }}</h1>
 <script type="module">
   import { createApp } from 'vue'
-  import i18n from './locale.js'
+  import i18nLocale from './locale.js'
```
```suggestion
  import i18n from './locale.js'
```

I'm confused why this was changed? I'm not seeing an i18nLocale in locale.js. Maybe it was just a bad copy pasta?

xanderfrangos replied:
Something changed with the localization library since I last worked on the PIN page months ago. The "success"/"failure" messages after pairing were throwing errors. 

The `i18n` import was previously an object with the function `i18n.global.t` (see [line 53](https://github.com/LizardByte/Sunshine/blob/bf8fdbb076207d69b7884491728f3c0f26fda1ac/src_assets/common/assets/web/pin.html#L53) for example). However, that `i18n` import is now itself an async function that returns the expected object (containing `global.t`). So I needed to await the import to get the object. This meant that I either needed to rename [the import](https://github.com/LizardByte/Sunshine/blob/bf8fdbb076207d69b7884491728f3c0f26fda1ac/src_assets/common/assets/web/pin.html#L28) or rename the [returned object](https://github.com/LizardByte/Sunshine/blob/bf8fdbb076207d69b7884491728f3c0f26fda1ac/src_assets/common/assets/web/pin.html#L40). I decided to rename the import (from `i18n` to `i18nLocale`) instead of having a new name for the returned object.

It technically works fine as-is. `locale.js` uses the async function as the default export, so it doesn't matter what we call the imported variable. If you prefer the import still be called `i18n`, we can rename [the returned object](https://github.com/LizardByte/Sunshine/blob/bf8fdbb076207d69b7884491728f3c0f26fda1ac/src_assets/common/assets/web/pin.html#L40) instead of the import. Or if there's a better way to access the localization library, we can do that.

ReenigneArcher replied:
@Hazer do you have any thoughts on this?

Hazer replied:
This way you're loading the locale twice, it does some API request behind, so it's not ideal.

When I created the `initApp` method, I already provided the `i18n` for injection, so you can inject it there

Hazer replied:
My suggestion rewriting like this:

```vue
<script type="module">
  import { createApp } from 'vue'
  import { initApp } from './init'
  import Navbar from './Navbar.vue'

  let app = createApp({
    components: {
      Navbar
    },
    inject: ['i18n'],
    methods: {
      registerDevice(e) {
        let pin = document.querySelector("#pin-input").value;
        let name = document.querySelector("#name-input").value;
        document.querySelector("#status").innerHTML = "";
        let b = JSON.stringify({pin: pin, name: name});
        fetch("/api/pin", {method: "POST", body: b})
          .then((response) => response.json())
          .then((response) => {
            if (response.status.toString().toLowerCase() === "true") {
              document.querySelector(
                "#status"
              ).innerHTML = `<div class="alert alert-success" role="alert">${this.i18n.t('pin.pair_success')}</div>`;
              document.querySelector("#pin-input").value = "";
              document.querySelector("#name-input").value = "";
            } else {
              document.querySelector(
                "#status"
              ).innerHTML = `<div class="alert alert-danger" role="alert">${this.i18n.t('pin.pair_failure')}</div>`;
            }
          });
      }
    }
  });

  initApp(app);
</script>
```

This way you can inject `i18n` easily, and also make a bit clearer what is this code.

Hazer replied:
Or, in case you're a bit adventurous, replace the whole `pin.html` with:

```vue
<!DOCTYPE html>
<html lang="en" data-bs-theme="auto">

<head>
  <%- header %>
</head>

<body id="app" v-cloak>
  <Navbar></Navbar>
  <div id="content" class="container">
    <h1 class="my-4 text-center">{{ $t('pin.pin_pairing') }}</h1>
    <pin-card></pin-card>
  </div>
</body>

<script type="module">
  import { createApp } from 'vue'
  import { initApp } from './init'
  import Navbar from './Navbar.vue'
  import PinCard from './PinCard.vue'

  let app = createApp({
    components: {
      Navbar,
      PinCard
    }
  });

  initApp(app);
</script>
```

and create a new file `PinCard.vue`:

```vue
<script setup>
import { ref } from 'vue'

const inputPIN = ref('')
const inputName = ref('')
const status = ref()

function registerDevice() {
  let b = JSON.stringify({pin: inputPIN.value, name: inputName.value});
  console.log(b)
  fetch("/api/pin", {method: "POST", body: b})
    .then((response) => response.json())
    .then((response) => {
      if (response.status.toString().toLowerCase() === "true") {
        status.value = true
        inputPIN.value = ""
        inputName.value = ""
      } else {
        status.value = false
      }
    });
}
</script>

<template>
  <form class="form d-flex flex-column align-items-center" id="form" @submit.prevent="registerDevice">
    <div class="card flex-column d-flex p-4 mb-4">
      <input
        type="text"
        pattern="\d*"
        v-model.trim="inputPIN"
        :placeholder="`${$t('navbar.pin')}`"
        id="pin-input"
        class="form-control mt-2"
        autofocus
        required />
      <input
        type="text"
        v-model.trim="inputName"
        :placeholder="`${$t('pin.device_name')}`"
        id="name-input"
        class="form-control my-4"
        required />
      <button class="btn btn-primary">{{ $t('pin.send') }}</button>
    </div>
    <div class="alert alert-warning">
      <b>{{ $t('_common.warning') }}</b> {{ $t('pin.warning_msg') }}
    </div>
    <div id="status" v-if="status !== undefined">
      <div class="alert alert-success" role="alert" v-if="status === true">{{ $t('pin.pair_success') }}</div>
      <div class="alert alert-danger" role="alert" v-else>{{ $t('pin.pair_failure') }}</div>
    </div>
  </form>
</template>

<style scoped>

</style>

```

xanderfrangos replied:
@ReenigneArcher I've updated it to @Hazer's code from the [latest comment](https://github.com/LizardByte/Sunshine/pull/2042#discussion_r1615940931). 👍

xanderfrangos commented:
> I don't know if there's a way to handle this without changes to Moonlight, but I did face this small issue.
> 
> When starting my testing initially, there was one unnamed device paired. I removed the Sunshine host from the client, and then repaired it. In Sunshine this showed 2 devices, one named and one unnamed.
> 
> When I removed the named one, it still had access since it was also the unnamed one.
> 
> Once, I removed the unnamed device on Sunshine's side, it was not able to connect.
> 
> Otherwise this looks good. I guess we can just recommend that people unpair all if they have any unnamed devices.

Yeah, there's not much that can be done about unnamed devices. There's currently no way to figure out which clients/devices they're for and properly name them (or check if they were already paired). But I expect it to be a short-lived issue, since setting a device name for new pairings is now required.

Hazer commented on src_assets/common/assets/web/pin.html:
```diff
@@ -8,10 +8,11 @@
   <div id="content" class="container">
-    <h1 class="my-4">{{ $t('pin.pin_pairing') }}</h1>
+    <h1 class="my-4 text-center">{{ $t('pin.pin_pairing') }}</h1>
     <form action="" class="form d-flex flex-column align-items-center" id="form">
```
With the change proposed below, use this:

```suggestion
    <form class="form d-flex flex-column align-items-center" id="form" @submit.prevent="registerDevice">
```

moi952 commented:
Hi,


Thanks for the dev :)
I tried development, and I only see unknown clients, couldn't we rename them as we connect with them?

Because I have a lot of unknown clients and I don't know which clients I'm using and which are old clients (potentially Android emulator, or old OS with the tests I've done)

Thanks



![image](https://github.com/LizardByte/Sunshine/assets/52632400/52e4f9cc-c83e-48dd-9928-b68a07b88940)

ReenigneArcher commented:
It's recommended to unpair all clients one time. Then you can name them as you pair them again.
